erigon-lib/state/history.go

```diff
@@ -643,53 +730,107 @@ func (h *History) collate(ctx context.Context, step, txFrom, txTo uint64, roTx k
 		defer cd.Close()
 	}
 
-	keyBuf := make([]byte, 0, 256)
-	for _, key := range keys {
-		bitmap := indexBitmaps[key]
-		it := bitmap.Iterator()
-		keyBuf = append(append(keyBuf[:0], []byte(key)...), make([]byte, 8)...)
-		lk := len([]byte(key))
+	efComp, err := seg.NewCompressor(ctx, "ef history", efHistoryPath, h.dirs.Tmp, seg.MinPatternScore, h.compressWorkers, log.LvlTrace, h.logger)
+	if err != nil {
+		return HistoryCollation{}, fmt.Errorf("create %s ef history compressor: %w", h.filenameBase, err)
+	}
+	if h.noFsync {
+		efComp.DisableFsync()
+	}
+	efHistoryComp = NewArchiveWriter(efComp, CompressNone)
+
+	bitmap := bitmapdb.NewBitmap64()
+	var prevKey, prevValue []byte
 
+	indexAddKV := func() error {
+		// fmt.Printf("indexAddKV(%v) | %x -> %x %p\n", bitmap.GetCardinality(), prevKey, k[:lk], bitmap)
+		ef := eliasfano32.NewEliasFano(bitmap.GetCardinality(), bitmap.Maximum())
+		it := bitmap.Iterator()
 		for it.HasNext() {
-			txNum := it.Next()
-			binary.BigEndian.PutUint64(keyBuf[lk:], txNum)
-			//TODO: use cursor range
-			if h.historyLargeValues {
-				val, err := roTx.GetOne(h.historyValsTable, keyBuf)
-				if err != nil {
-					return HistoryCollation{}, fmt.Errorf("getBeforeTxNum %s history val [%x]: %w", h.filenameBase, key, err)
-				}
-				if len(val) == 0 {
-					val = nil
-				}
-				if err = historyComp.AddWord(val); err != nil {
-					return HistoryCollation{}, fmt.Errorf("add %s history val [%x]=>[%x]: %w", h.filenameBase, key, val, err)
-				}
+			ef.AddOffset(it.Next())
+		}
+		bitmap.Clear()
+		ef.Build()
+
+		prevValue = ef.AppendBytes(prevValue[:0])
+
+		if err = efHistoryComp.AddWord(prevKey); err != nil {
```

Comment:
are you sure we not loose last key here? seems we call `indexAddKV()` only in 1 place